### ISSUE TITLE
"Error: Corrupted zip or bug: expected 6409 records in central dir, got 0" with small zip

### ISSUE THREAD

**Issue by hugovk** [dariusk/twitter-archiver#10]
Using a recent Twitter archive zip (for the recently deleted [@PythonPillow](https://web.archive.org/web/20240212050203/https://twitter.com/PythonPillow)), "Your archive.html" says: 

> Date generated·February 27, 2024 at 6:39:12 PM GMT+2·Estimated size·11 MB

The zip is actually 19 MB which is still small. It has 190 tweets.

I got this:

> **Error! Error: Corrupted zip or bug: expected 6409 records in central dir, got 0**
> 
> I am guessing that your archive is too big! If it's more than 2GB you're likely to see this error. If you look above at the "Known problems" section, you'll see a potential solution. Sorry it is a bit of a pain in the ass.

I fixed it by unzipping it and re-zipping it (on macOS with right-click the contents and "Compress"), and it worked with that file.

---

PS thank you for this tool! The archive is now hosted at https://python-pillow.org/twitter-archive/

**Comment by mrcolbyrussell** [comment]
@hugovk, do you still have the original archive?  Can you provide a test case?

**Comment by mrcolbyrussell** [comment]
@aclark4life, same question to you, so we can maybe fix this for other people.

re <https://github.com/python-pillow/Pillow/issues/7828#issuecomment-2031953696>

> I couldn't get https://tinysubversions.com/twitter-archive/make-your-own/ to work but if anyone else wants to try please let me know and I'll send you the zip. Pretty sure I've published the contents of the zip on the web in the past without worrying about privacy issues

**Comment by aclark4life** [comment]
@mrcolbyrussell I have the original and can get new archives from Twitter for my personal account but hesitant to share due to privacy issues. However if you have PRs or branches with potential fixes I can test for you. Hope that helps.

**Comment by mrcolbyrussell** [comment]
@aclark4life I understand the reluctance about privacy issues for your personal account.  I'm fine just using with the PythonPillow archive.

> if you have PRs or branches with potential fixes I can test for you

I don't have anything.  I have no way to even use this; I don't have a ZIP to feed into it.

**Comment by gamebits** [comment]
@mrcolbyrussell I have a ZIP file that's experiencing this issue and which I can provide for testing.

**Comment by mrcolbyrussell** [comment]
@gamebits whence?

**Comment by gamebits** [comment]
@mrcolbyrussell Is there a way I can submit it to you other than posting it publicly here?

**Comment by mrcolbyrussell** [comment]
@gamebit how big is the archive?  You can email it to me if it's small enough.  My contact info is in my profile and my homepage. [<https://colbyrussell.com/>](https://colbyrussell.com/)

If it's too big for email and you want to upload to a filesharing service like Google Drive, you could do that and email me the link—just make sure you set the right permissions.  You could of course also fork/create a private GitHub project (maybe github.com/gamebits/twitter-archive) and add me as a collaborator.  Flip over to the releases tab from the repo overview and upload the ZIP as a binary.  (For good measure, you could also unzip the archive into a Git repo and push it to the same project, too.  Regardless of whether you do this, you'll still want the actual ZIP that's causing problems to be uploaded, though, since we're interested in the file as provided by Twitter, and it won't do me any good to clone a copy that's already been unzipped—having the repo would just be if you felt like using Git/GitHub to manage your backups was a good idea.)

**Comment by waldyrious** [comment]
I faced exactly the same issue. I have emailed both the original zip file exported from twitter, and the unzipped-rezipped version to @mrcolbyrussell's email address as indicated in his website. Happy to provide the links in case anyone else would like to investigate as well.

**Comment by larsschwarz** [comment]
Make sure the /data directory is on root level of the ZIP directory structure. So if you unzip your zip it should look like (w/o any parent directory like the name of your archive):

```
assets
Your archive.html
data
```
For example: The original Twitter/X archive download is entitled: 

twitter-2025-02-04-445a906f97d9b05f195a653f3cfdb8bc135680a0427c630c6b9bd6b4149e8cdf.zip

If you unzip this it'll include the filename as root directory name. You don't want that. 

Tested this with an archive downloaded today and it works just fine (saw the same errors before when I tried to provide the original ZIP or a re-zipped version).

**Comment by mrcolbyrussell** [comment]
@larsschwarz directory nesting level has nothing to do with the error that is the subject of this bug.

Thanks to the archive provided by @waldyrious, we've determined that the actual culprit here is poor (i.e. mostly if not totally broken) support for Zip64 in the jszip dependency.  This is [jszip issue #922](https://github.com/Stuk/jszip/issues/922).  The jszip package, for whatever reason, expects zero bytes between the last file entry in the central directory and the four bytes comprising the magic central directory signature.  This is a bad assumption considering a) it has no bearing whatsoever on being able to work with the data in the archive, and b) the Zip64 locator appearing after the final file entry is exactly where the diagram in the spec indicates the locator should be.

What's worse, though, is that it's not just that the locator is there and jszip assumes that it won't be, but the fatal error is caused by the bad, bad way that jszip tries to "work around" it being there.  (To reiterate: there is absolutely no workaround required—and that goes for implementations that advertise *no* support for Zip64!)

(To be fair, poking around the internals of the archive that Twitter spits out reveals some oddities—like the fact that it's using Zip64 at all in cases where it isn't necessary and "classic" ZIP would suffice, even with its limitation on file sizes and numbers of archived files.  This isn't exculpatory of jszip's brokenness, though.)

**Comment by mrcolbyrussell** [comment]
While I'm here and since GitHub has put up so many barriers/obstructions (like unwanted and unwelcome 2FA requirements) for someone trying to log in and leave a comment:

I recommend @dariusk ditch jszip and use something like @greggman's [unzipit](https://github.com/greggman/unzipit) for ZIP support.  It is better than jszip in a number of ways, including the fact that it is smaller, more correct to my knowledge, and doesn't have the same inherent performance problems.  (The code is also *much* cleaner than that found in the jszip blob included in the repo here.)

Selection/"upload" of directories (rather than single files) is also well-supported by every major browser released in the last 5 years or so.  (The last holdout on this was Safari; despite this feature being activated by [the `webkitdirectory` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#webkitdirectory), for the longest time it was only Firefox and Chrome that supported this.  The origins of the attribute name date back to when Chrome/Chromium was based on Webkit and got built with a bunch of Chrome-specific feature flags enabled, which controlled everything from whether or not it would support `webkitdirectory` in the browser filepicker to which JS engine to use—JavaScriptCore or V8.)

**Comment by mrcolbyrussell** [comment]
For those trying to use this project, note also something that I pointed out to @waldyrious in private correspondence, which is that the archive that Twitter gives you includes its own (very nice) data viewer (in the root of the archive—all you have to do is unzip it and double-click the `Your archive.html` "app", which should automatically open in your browser on almost any system).

If your main motivation is to be able to browse/view your tweets on your own computer rather than converting them into a public archive of static HTML that you can host on your own site, then for this reason (along with issue #12), you might consider just unzipping your archive and using the bundled viewer rather than running it through a tool like this one.

**Comment by waldyrious** [comment]
Thanks for the thorough investigation and troubleshooting!

> Selection/"upload" of directories (rather than single files) is also well-supported by every major browser released in the last 5 years or so.

Do you mean that, as an alternative to replacing jszip with unzipit, it might be possible to support unzipped archives, so that one should be able to unzip the archive from Twitter and upload the resulting directory directly to the archiver, without re-zipping it?

**Comment by waldyrious** [comment]
> note also something that I pointed out to @waldyrious in private correspondence, which is that the archive that Twitter gives you includes its own (very nice) data viewer (in the root of the archive—all you have to do is unzip it and double-click the `Your archive.html` "app", which should automatically open in your browser on almost any system).

For reference, here's a comparison of what browsing my tweets looks like using this archive vs. the embedded viewer:

| dariusk/twitter-archiver | Twitter's `Your archive.html` |
|-|-|
| ![Image](https://github.com/user-attachments/assets/82d49635-0d06-48cc-abcc-1964571e8abe) | ![Image](https://github.com/user-attachments/assets/6f614b00-395b-4a52-b815-8c6d00d11dd5) |

Of course, the former is certainly more hackable and stylable, includes separate linkable HTML files for each tweet, etc. Both rely on javascript to run, though. Also, Twitter's native archive viewer exposes a lot of private data, including account details, DMs, etc.

**Comment by mrcolbyrussell** [comment]
> Do you mean that, as an alternative to replacing jszip with unzipit, it might be possible to support unzipped archives, so that one should be able to unzip the archive from Twitter and upload the resulting directory directly

@waldyrious yes.  I have not tested, but this may/should also alleviate the problem of very large archives that is mentioned in the disclaimers and that @dariusk and I discussed on the fediverse a while back.

(In principle, there's nothing that requires browser developers implement `webkitdirectory` support in such a way that it requires putting the full contents into memory.  This is unlike the way that very naive ZIP processors on big data sets will do things.  I also noticed even with the relatively small (<100MB) archive you provided that feeding it into this tool led to noticeable performance problems, including high memory use and long processing times.)
